src/transport/SecureSessionMgr.cpp

```diff
@@ -110,55 +124,94 @@ CHIP_ERROR SecureSessionMgr::SendMessage(PayloadHeader & payloadHeader, NodeId p
     mPeerConnections.MarkConnectionActive(state);
 
     {
-        uint8_t * data = nullptr;
+        uint8_t * data         = nullptr;
+        uint8_t * p            = nullptr;
+        uint32_t msgId         = 0;
+        uint32_t payloadLength = 0;
+        uint16_t len           = 0;
         PacketHeader packetHeader;
         MessageAuthenticationCode mac;
 
         const uint16_t headerSize = payloadHeader.EncodeSizeBytes();
         uint16_t actualEncodedHeaderSize;
         uint16_t totalLen = 0;
         uint16_t taglen   = 0;
-        uint32_t payloadLength; // Make sure it's big enough to add two 16-bit
-                                // ints without overflowing.
-        static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
-                      "Addition to generate payloadLength might overflow");
-        payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
-        VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+
+        if (!isEncrpted)
+        {
+            msgId = state->GetSendMessageIndex();
+
+            static_assert(std::is_same<decltype(msgBuf->TotalLength()), uint16_t>::value,
+                          "Addition to generate payloadLength might overflow");
+            payloadLength = static_cast<uint32_t>(headerSize + msgBuf->TotalLength());
+            VerifyOrExit(CanCastTo<uint16_t>(payloadLength), err = CHIP_ERROR_NO_MEMORY);
+        }
+        else
+        {
+            msgId         = bufferRetainSlot->mMsgId;
+            payloadLength = bufferRetainSlot->mPayloadLen;
```

Comment:
This is a bit weird.  The caller does `std::move` out of the slot in the retransmit table, but here we are relying on that slot still having the right message id and payload length, when conceptually it's garbage data that's been moved out of.  It would make more sense to have those be arguments to this method (which are only examined if `isEncrypted` is true) and get them from the incoming `EncryptedPacketBufferHandle` in the corresponding `SendMessage` overload....

Reply:
Acked